docs/development/logging.md

```diff
@@ -0,0 +1,165 @@
+Hence, we are migrating our component logging from logrus to logr (backed by zap) as part of [gardener/gardener#4251](https://github.com/gardener/gardener/issues/4251).
+
+> ⚠️ `logger.Logger` (logrus logger) is deprecated in Gardener and shall not be used in new code – use logr loggers when writing new code!
+> 
+> ℹ️ Don't use zap loggers directly, always use the logr interface in order to avoid tight coupling to a specific logging implementation.
+
+gardener-apiserver differs from the other components as it is based on the [apiserver library](https://github.com/kubernetes/apiserver) and therefore uses [klog](https://github.com/kubernetes/klog) – just like kube-apiserver.
+As gardener-apiserver writes (almost) no logs in our coding (outside of the apiserver library), there is currently no plan for switching the logging implementation.
+Hence, the following sections focus on logging in the controller and admission components only.
+
+## Structured Logging
+
+Similar to [efforts in the Kubernetes project](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md), we want to migrate our component logs to structured logging.
+As motivated above, we will use the logr interface instead of klog though.
+
+You can read more about the motivation behind structured logging in [logr's background and FAQ](https://github.com/go-logr/logr#background) (also see [this blog post by Dave Cheney](http://dave.cheney.net/2015/11/05/lets-talk-about-logging)).
+Also, make sure to check out controller-runtime's [logging guideline](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.11.0/TMP-LOGGING.md) with specifics for projects using the library.
+The following sections will focus on the most important takeaways from those guidelines and give general instructions on how to apply them to Gardener and its controller-runtime components.
+Note: some parts in this guideline differ slightly from controller-runtime's document.
+
+### TL;DR of Structured Logging
+
+❌ stop using `printf`-style logging:
+```go
+var logger *logrus.Logger
+logger.Infof("scaling deployment %s/%s to %d replicas", deployment.Namespace, deployment.Name, replicaCount)
+```
+
+✅ instead, write static log messages and enrich them with additional structured information in form of key-value pairs:
+
+```go
+var logger logr.Logger
+logger.Info("scaling deployment", "deployment", client.ObjectKeyFromObject(deployment), "replicas", replicaCount)
+```
+
+## Log Configuration
+
+Gardener components can be configured to either log in `json` (default) or `text` format:
+`json` format is supposed to be used in production, while `text` format might be nicer for development.
```

Comment:
For the local setup, shouldn't we unify all components (except gardener-apiserver probably) to log in `json`? I guess today gardenlet is configured to log in `json`, while GCM is still on `text`.

Reply:
Hmm, in `example/20-componentconfig-{gardener-controller-manager,gardener-scheduler,gardenlet}`, `logFormat` is set to `text`. So it's already unified, right?

Reply:
You are right, my gardenlet component config under `dev/` was outdated.